crit-go/images/images.go

```diff
@@ -0,0 +1,240 @@
+	magicjson()
+	m := magichandler(f)
+	image["magic"] = m
+	switch {
+	case m == "GHOST_FILE":
+		handler := &ghost_file_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "PAGEMAP":
+		handler := &pagemap_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "PIPES_DATA":
+		handler := &pipes_data_extra_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "FIFO_DATA":
+		handler := &pipes_data_extra_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "SK_QUEUES":
+		handler := &sk_queues_extra_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "IPCNS_SHM":
+		handler := &ipc_shm_set_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "IPCNS_SEM":
+		handler := &ipc_sem_set_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "IPCNS_MSG":
+		handler := &ipc_msg_queue_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	case m == "TCP_STREAM":
+		handler := &tcp_stream_extra_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	default:
+		handler := &entry_handler{}
+		handler.m = m
+		image["entries"] = handler.Load(f, pretty, nopl)
+	}
+	f.Close()
+	return image
+}
+
+func Dump(ilf *os.File, ouf *os.File) {
+	/*
+	   Convert criu image from dict(json) format to binary.
+	   Takes an image in dict(json) format and file-like
+	   object to write to.
+	*/
+	magicjson()
+	var img map[string]interface{}
+	file, err := ioutil.ReadAll(ilf)
+	if err != nil {
+		fmt.Println("Error reading Input Json file: %s", err)
+		ilf.Close()
+		ouf.Close()
+		os.Exit(1)
+	}
+	if err := json.Unmarshal(file, &img); err != nil {
+		fmt.Println("Error Unmarshalling Json file : %s", err)
+		ilf.Close()
+		ouf.Close()
+		os.Exit(1)
+	}
+	ilf.Close()
+	m := img["magic"].(string)
+	magic_val, err := strconv.ParseUint(bynamemap[m].(string), 10, 64)
+	if err != nil {
+		fmt.Println("Unable to Parse magic value from file")
+		checkfile(err, ouf)
+	}
+	if m != "INVENTORY" {
+		if m == "STATS" || m == "IRMAP_CACHE" {
+			bnvmap, err := strconv.ParseUint(bynamemap["IMG_SERVICE"].(string), 10, 64)
+			checkfile(err, ouf)
+			bs := make([]byte, 4)
+			binary.LittleEndian.PutUint32(bs, uint32(bnvmap))
+			_, err = ouf.Write(bs)
+			checkfile(err, ouf)
+		} else {
+			_, err := strconv.ParseUint(bynamemap["IMG_COMMON"].(string), 10, 64) //.(string) //.(float64)
+			checkfile(err, ouf)
+			bnvmap, err := strconv.ParseUint(bynamemap["IMG_COMMON"].(string), 10, 64)
+			checkfile(err, ouf)
+			bs := make([]byte, 4)
+			binary.LittleEndian.PutUint32(bs, uint32(bnvmap))
+			_, err = ouf.Write(bs)
+			checkfile(err, ouf)
+		}
+	}
+	bs := make([]byte, 4)
+	binary.LittleEndian.PutUint32(bs, uint32(magic_val))
+	_, err = ouf.Write(bs)
+	if err != nil {
+		fmt.Println(err)
```

Comment:
it has to exit with a non-zero code if it failed to write into the file.

Reply:
In some places i had to skip the exit. Example : [L233](https://github.com/Pac23/go-criu/blob/59082e70b35a0b2d295de9a7435fbc735d0a4d31/crit-go/images/handle.go#L233) jsonpb complains about the extra data not being a field in the protobinding but will output the .img file. So Exit had to be avoided in all the dump functions of the extra handlers.